Add the ability to record / playback multiple versions

m-nash opened the pull request:
To support AzureStack we want to be able to certify the SDK with multiple versions without needing to rewrite test cases.  There was already a mechanism that allowed for multiple versions in live mode [here ](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/core/Azure.Core.TestFramework#support-multi-service-version-testing).  This PR extends that functionality in a few ways.

1. Allows for the version to be either an Enum convertible to Int (previous behavior) OR a string (new behavior)
2. Adds one new parameter to the `ClientTestFixture` constructor to indicate if you want the versions to apply to record / playback (default being false which maintains previous behavior)
3. Adds a new qualifier to the filename to differentiate between which recording belongs to which version

# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

m-nash committed:
wip for testing the ability to record / playback multiple versions

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[Azure.ResourceManager](https://apiview.dev/Assemblies/Review/d443d938a20a436abb8795bf29dba243)

JoshLove-msft commented on sdk/core/Azure.Core.TestFramework/src/ClientTestFixtureAttribute.cs:
```diff
@@ -40,15 +42,23 @@ public class ClientTestFixtureAttribute : NUnitAttribute, IFixtureBuilder2, IPre
         /// <param name="serviceVersions">The set of service versions that will be passed to the test suite.</param>
         /// <param name="additionalParameters">An array of additional parameters that will be passed to the test suite.</param>
-        public ClientTestFixtureAttribute(object[] serviceVersions, object[] additionalParameters)
+        public ClientTestFixtureAttribute(object[] serviceVersions, object[] additionalParameters, bool recordMultipleVersions = false)
```
nit: we chatted about something like `recordAllVersions`

JoshLove-msft commented:
Overall approach looks good. Added a couple comments for things we discussed offline.
Here are the list of the test classes/patterns that can be moved to a new Azure.Core.TestFramework.Tests project that @m-nash offered to create:
- ClientTestBase*
- RecordedTestBaseTests
- TestEnvironmentTests
- ManagementRecordedTestBaseTests
- IgnoreServiceErrorAttributeTests

m-nash committed:
address feedback and normalize version types

m-nash committed:
update location of test project
move clientTestBase* tests to new test project
few minor fixes on clienttestfixtureattribute

m-nash committed:
move recordedtestbase and testenvironmenttests

m-nash committed:
create shared source for test framework helpers and move mgmtrecordedtestbasetests

m-nash committed:
move recordings

m-nash committed:
merge from main

m-nash committed:
update after merge

m-nash committed:
update namespaces

m-nash committed:
fix a couple issues

m-nash committed:
regen ResourceGroupResource tests for multi-version

m-nash committed:
remove unused file

JoshLove-msft commented on sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:
```diff
@@ -177,9 +179,10 @@ public string ReplacementHost
         };
 
-        protected RecordedTestBase(bool isAsync, RecordedTestMode? mode = null) : base(isAsync)
+        protected RecordedTestBase(bool isAsync, RecordedTestMode? mode = null, string apiVersion = null) : base(isAsync)
```
nit: call the param versionQualifier?

JoshLove-msft commented on sdk/core/Azure.Core.TestFramework/src/ClientTestFixtureAttribute.cs:
```diff
@@ -15,13 +16,49 @@ namespace Azure.Core.TestFramework
     {
+        private readonly struct TestVersion : IComparable
+        {
+            public IComparable Comparable { get; }
```
Do we need this property? Can't we just store the converted data in `Object` and use CompareTo on that.

m-nash replied:
Object is actually the only thing needed outside this struct on line 169, but I can make this a private readonly field

JoshLove-msft commented on sdk/core/Azure.Core.TestFramework/src/ClientTestFixtureAttribute.cs:
```diff
@@ -37,18 +74,32 @@ public class ClientTestFixtureAttribute : NUnitAttribute, IFixtureBuilder2, IPre
         /// <param name="serviceVersions">The set of service versions that will be passed to the test suite.</param>
         /// <param name="additionalParameters">An array of additional parameters that will be passed to the test suite.</param>
-        public ClientTestFixtureAttribute(object[] serviceVersions, object[] additionalParameters)
+        public ClientTestFixtureAttribute(object[] serviceVersions, object[] additionalParameters, bool recordAllVersions = false)
```
nit: add param docs

JoshLove-msft commented on sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:
```diff
@@ -177,9 +179,10 @@ public string ReplacementHost
         };
 
-        protected RecordedTestBase(bool isAsync, RecordedTestMode? mode = null) : base(isAsync)
+        protected RecordedTestBase(bool isAsync, RecordedTestMode? mode = null, string apiVersion = null) : base(isAsync)
```
We should also document this constructor and parameters explaining the usage and that it is only currently used for mgmt plane.

christothes replied:
Should we also consider creating a separate ctor overload that takes this argument, just to make it less visible to data plane test classes?

m-nash replied:
This is not intended to be a mgmt plane feature, it allows you record / playback specific versions where as before the only functionality was to be able to run live tests on multiple versions.  Which SDKs opt in or out today shouldn't sway our design since that will change over time.

m-nash replied:
This was resolved when we were fixing a consistency issue between two settings so resolving as no-op at this point.

JoshLove-msft commented on sdk/core/Azure.Core.TestFramework/src/VersionString.cs:
```diff
@@ -0,0 +1,81 @@
+            IsGa = true;
+            int index = -1;
+            PreviewString = null;
+            if (chars.Length > 10)
```
Why 10?

christothes replied:
I think that is the standard length for a version string minus any preview suffixes. `nnnn-nn-nn`

JoshLove-msft replied:
Yes, that is what I assumed, but I am hinting at the need for a constant or a comment.

JoshLove-msft commented on sdk/core/Azure.Core.TestFramework/src/VersionString.cs:
```diff
@@ -0,0 +1,81 @@
+
+namespace Azure.Core.TestFramework
+{
+    internal readonly struct VersionString : IComparable
```
Is this still needed?

m-nash replied:
I need somewhere to encaspulate the parsing / comparison of a version string

JoshLove-msft replied:
Can we move it to where it is used as part of TestVersion?

m-nash replied:
Its cleaner to separate this as its own thing.  There is also potential here to share this logic since this isn't the only place we do this, but I am punting on that for now I have a follow up I will track

christothes commented on sdk/core/Azure.Core.TestFramework/src/ClientTestFixtureAttribute.cs:
```diff
@@ -37,18 +74,32 @@ public class ClientTestFixtureAttribute : NUnitAttribute, IFixtureBuilder2, IPre
+            {
+                var first = _serviceVersions[0];
+                if (!_serviceVersions.All(x => x.GetType() == first.GetType()))
+                    throw new InvalidOperationException("All service versions must be the same type");
```
should we use the same string as above? 
"The items in the serviceVersions array must be an enum convertable to Int32 or a string in date format with an optional preview string"

m-nash replied:
This error is specifically testing for homogenousness, but I can add an additional part which specifies the valid types

m-nash committed:
Merge branch 'main' of https://github.com/Azure/azure-sdk-for-net into mnash-testframeworkMultiVersion

m-nash committed:
address comments

JoshLove-msft commented on sdk/core/Azure.Core.TestFramework/src/ClientTestFixtureAttribute.cs:
```diff
@@ -18,18 +18,18 @@ public class ClientTestFixtureAttribute : NUnitAttribute, IFixtureBuilder2, IPre
         private readonly struct TestVersion : IComparable
         {
-            public IComparable Comparable { get; }
+            private readonly IComparable _comparable;
```
Still don't see why we need this at all, as compared to just storing in Object.

m-nash replied:
Object is the original thing that was passed in which is needed on line 170 (used to be 169 in a previous comment but it moved).  _comparable is the "converted thing" that can be used to compare for finding the Min / Max which is of a completely different type.

If I don't keep both in this struct then I will need some sort of dictionary to look up the original object from the comparable after the Max method is called.

JoshLove-msft replied:
Ah, got it.

m-nash committed:
initial draft on readme changes

m-nash committed:
address consistency between two flags and collapse into one

m-nash committed:
minor update on readme

JoshLove-msft reviewed (approved):
LGTM!
